### PR TITLE
Persist deduplication strategy in url

### LINKED ISSUE #14445: Explore: Logs dedup strategy is not included in url state
When refreshing page or sending link the dedup strategy option is lost.

### PR DESCRIPTION
**Pull request by dprokop**
Fixes #14445 

- Moved dedup strategy state out from Logs to redux store.
- Introduced thunk for persisting UI state in explore `updateExploreUIState`. It dispatches     `updateUIStateAction` and saves state to URL. Reducer trivially updates Explore's state fragment responsible for UI (showLogs, showGraph, showTable, deduplicationStrategy). Not happy with this reducer at all, but these UI changes are trivial and binary ¯\_(ツ)_/¯ .
- Updated `toggle[Graph|Logs|Table]` to dispatch `updateExploreUIState` instead of manually updating UI state in reducer and saving state to URL.

### CODE REVIEW

**Comment by dprokop** [comment]
WDYT about moving deduplicated logs calculation to redux? `Logs` component does a lot in the render function.

**Comment by hugohaggmark** [comment]
On my last job we did all data processing in reducers, why?
++ One single point of truth
++ Easily tested and debugged with redux-logger
++ The components became very simple

-- Components rely on state or "Container Components" to work

I think we do more stateful components here, maybe we should talk about the roadmap/goal for redux/state in the team, WDYT?

**Comment by dprokop** [comment]
I'm towards this approach as well @hugohaggmark. It's not only easier to test such code, but also gives a clear guideline to everyone involved in development, where this and that piece of code should belong. 

> -- Components rely on state or "Container Components" to work
I understand this concern. But on the other hand this approach enables the UI development before the domain logic work has even started. Also, if we take test / storybook / visual regression tests POV into consideration, then having these dumb components rendered without the whole redux ceremony becomes trivial and is just a matter of correct mocks.

**Comment by davkal** [comment]
Works well.

> moving deduplicated logs calculation to redux?

Dedup currently happens after level filtering, which may lower dedup efforts considerably. Unless level filtering moves to redux too, dedup will always have to be done on the whole set. Dedup currently takes 500ms on 1000 lines on my MBA. But it's not optimal to have that much work be done in `render`.

File: public/app/features/explore/LogsContainer.tsx [hunk 1]
```diff
@@ -69,11 +77,23 @@ export class LogsContainer extends PureComponent<LogsContainerProps> {
...
+  };
+};
 function mapStateToProps(state: StoreState, { exploreId }) {
   const explore = state.explore;
   const item: ExploreItemState = explore[exploreId];
-  const { logsHighlighterExpressions, logsResult, queryTransactions, scanning, scanRange, showingLogs, range } = item;
+  const { logsHighlighterExpressions, logsResult, queryTransactions, scanning, scanRange, range } = item;
   const loading = queryTransactions.some(qt => qt.resultType === 'Logs' && !qt.done);
+  const {showingLogs, dedupStrategy} = selectItemUIState(item);
+  // const dedup = item.dedup;
```
Comment (davkal) [inline public/app/features/explore/LogsContainer.tsx]: Ooops

File: public/app/features/explore/state/actionTypes.ts [hunk 2]
```diff
@@ -180,6 +180,8 @@ export interface SplitOpenPayload {
   itemState: ExploreItemState;
 }
 
+//
```
Comment (davkal) [inline public/app/features/explore/state/actionTypes.ts]: Ooops

File: public/app/features/explore/state/actionTypes.ts [hunk 3]
```diff
@@ -366,6 +372,11 @@ export const splitCloseAction = noPayloadActionCreatorFactory('explore/SPLIT_CLO
 export const splitOpenAction = actionCreatorFactory<SplitOpenPayload>('explore/SPLIT_OPEN').create();
 export const stateSaveAction = noPayloadActionCreatorFactory('explore/STATE_SAVE').create();
 
+/**
+ * Update state of Explores UI
```
Comment (davkal) [inline public/app/features/explore/state/actionTypes.ts]: This is a bit too general. Which part of the state? Where is it kept? Is it all state?

**Comment by dprokop** [comment]
@davkal Thanks for the review!

> Dedup currently happens after level filtering, which may lower dedup efforts considerably. Unless level filtering moves to redux too, dedup will always have to be done on the whole set. Dedup currently takes 500ms on 1000 lines on my MBA. But it's not optimal to have that much work be done in `render`.

Yes, I'm aware of that, and moving filtering to redux is a part of the effort of migrating dedup calc to redux as well.

File: public/app/features/explore/state/actions.ts [hunk 4]
```diff
@@ -698,22 +716,29 @@ const togglePanelActionCreator = (
...
-    dispatch(actionCreator({ exploreId }));
-    dispatch(stateSave());
+    let shouldRunQueries, uiFragmentStateUpdate: Partial<ExploreUIState>;
 
     switch (actionCreator.type) {
       case toggleGraphAction.type:
-        shouldRunQueries = getState().explore[exploreId].showingGraph;
+        const isShowingGraph = getState().explore[exploreId].showingGraph;
+        shouldRunQueries = !isShowingGraph;
+        uiFragmentStateUpdate = { showingGraph: !isShowingGraph };
```
Comment (davkal) [inline public/app/features/explore/state/actions.ts]: Always a bit annoying if the action creators need to know the state shape.

Comment (dprokop) [reply public/app/features/explore/state/actions.ts]: Agree, this should be a part of the part of the payload.

Comment (dprokop) [reply public/app/features/explore/state/actions.ts]: @davkal Changes in #cee2e47

**Pull request merged**